### DISCUSSION TITLE
simulating array with multiple orientations within a string

### DISCUSSION THREAD

**Discussion by planted-jason-hoving** [pvlib/pvlib-python#2265] [Q&A]
hello, I am new to `pvlib` but loving the flexibility and granularity it provides!

I can see from examples (https://pvlib-python.readthedocs.io/en/stable/gallery/irradiance-transposition/plot_mixed_orientation.html) and several discussions/issues how it's possible to simulate arrays with multiple orientations **in parallel** (separate strings) but wondering if it is possible to chain multiple arrays **in series**.

The resulting IV curve mismatch is very detrimental to performance, similar to shading effects, so this is typically avoided in practice. However, there are two reasons for why I was hoping to model this:

  1. Illustrate the power loss of chaining east and west facing modules in series (in a single string)
  2. Study the impacts of slight perturbations in orientation due to installation tolerances. Modules within a string would be the same nominal orientation but could be perturbed by a few degrees

Maybe this could be done by artificially extracting IV curves to manipulate and combine manually? Seems like it would be something similar to what was done for partial shading of a single module here: https://pvlib-python.readthedocs.io/en/stable/gallery/shading/plot_partial_module_shading_simple.html

**Comment by kurt-rhee** [comment]
I am not a maintainer so please take my comments with a grain of salt.  I don't think that this exists today in pvlib, though there has been discussion about including functionality for this from a few different sources.

One option for doing this would be to use https://github.com/SunPower/PVMismatch.

**Comment by markcampanelli** [comment]
Not sure if this will shed any light on your computations: https://github.com/pvlib/pvlib-python/pull/1923

**Comment by cwhanse** [comment]
pvlib does not currently offer a calculator for the IV curve of series-connected mismatched strings. It is a high-interest development item.

**Comment by planted-jason-hoving** [comment]
Thanks all for the info! I tried out @markcampanelli 's code as well as some of my own that took concepts from the partial shading exercise.

I also got pretty good agreement when trying to fit a SDE approximation to the resulting IV curve (via `ivtools.sde.fit_sandia_simple`). Before I start using that approximation I just wanted to double check: in addition to no IV curve calculators, there is currently no way to simulate series-connected mismatched strings either right? Simulate meaning build with Array/PVSystem/ModelChain and run a weather profile.

![image](https://github.com/user-attachments/assets/9658b612-10fa-4a89-b0c1-a80a16057a21)
![image](https://github.com/user-attachments/assets/93b8e3f1-12d9-4c7e-bf5e-41080d475417)
![image](https://github.com/user-attachments/assets/553a6c82-e731-4c48-a715-ff495744295a)

**Comment by cwhanse** [comment]
pvlib has single diode equations solvers, see pvlib.pvsystem.singlediode and pvlib.pvsystem.i_from_v. These calculators assume that every cell in a series-connected string has the same irradiance and temperature. pvlib does not currently have a function to simulate a series-connected mismatched string.

The IV curves above look correct, since the calculation appears to assume a breakdown voltage less than -100V (off the scale to the left). The stepped shape of a mismatch IV curve results when the lower current string's reverse bias reaches the breakdown voltage, causing that string to conduct current much greater than its Isc. That's not happening in your simulation. Consequently the single diode equation fits well, which is what fit_sandia_simple is doing for you.
